core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotationUtil.java

```diff
@@ -70,11 +77,17 @@
     }
 
     /**
-     * Built-in meta annotations defined in {@code java.lang.annotation} package.
+     * A thread-local hash set of annotation classes that contain cyclic references.
      */
-    private static final Set<Class<? extends Annotation>> BUILT_IN_META_ANNOTATIONS =
-            ImmutableSet.of(Documented.class, Inherited.class, Native.class,
-                            Retention.class, Repeatable.class, Target.class);
+    private static final ThreadLocal<Set<Class<? extends Annotation>>> knownCyclicAnnotationTypes =
```

Comment:
Chose to use a thread-local variable instead of a global one to avoid class loader leak.

Reply:
Actually I have to use `WeakHashMap` to avoid class loader leak. Not enough sleep :-)

Reply:
Mostly out of curiosity, if using a weak map, do you still need to use threadlocal?

Reply:
I thought `WeakHashMap` is not thread safe, no?

Reply:
.. or did you mean using a concurrent version of weak key hash map?

Reply:
Yeah was comparing to using synchronizedMap with a weak hash map.

Reply:
Yeah, switched to `MapMaker`.